### ISSUE TITLE
iOS example audioInputExample: issue with audioIn / ofSoundPlayer order

### ISSUE THREAD

**Issue by dimitre** [openframeworks/openFrameworks#7291]
a bit more investigation: if the ofSoundStreamSetup() comes after the sound.load(), the microphone "wins" (audio input / no sound player). so it seems there is a conflict in the AVSession setups -- probably unrelated to the memory management, so perhaps this should be merged, and figure out the conflict in another PR?

_Originally posted by @artificiel in https://github.com/openframeworks/openFrameworks/issues/7283#issuecomment-1409743155_

It seems something in soundPlayer changes the way ofSoundStream behaves.

**Comment by artificiel** [comment]
indeed, the sound does work out of the headphones... probably related to the way each class interacts with `[AVAudioSession sharedInstance]`... 

https://github.com/openframeworks/openFrameworks/blob/625df269a462dcbc988ca0bcd90aafd0b66767bf/addons/ofxiOS/src/sound/SoundStream.m#L77

vs

https://github.com/openframeworks/openFrameworks/blob/625df269a462dcbc988ca0bcd90aafd0b66767bf/addons/ofxiOS/src/sound/AVSoundPlayer.m#L27

I tried adding `AVAudioSessionCategoryOptionDefaultToSpeaker` to both, with no difference. dealing with AVAudioSession and automatic audio routing seems tricky (lots of not-clearly-answered questions on stackoverflow). also no effect with:

```objc
[audioSession overrideOutputAudioPort:AVAudioSessionPortOverrideSpeaker
                                error:&audioSessionError];
```

the fact that the order of calls bet wound sound.load() and ofSoundStreamSetup() changes the behaviour feels like a bug, but is it really? or is it simply the result of the complicated (and potentially evolving) iOS behaviour for which it is difficult to provide an homogeneous OF-style interface? 

I guess there is a problem if OF wants to expose all the iOS routing possibilities (ex: airplay). as such, it would be good to be able to switch the sound routing from an OF interface. but maybe it's beyond the scope of OF, and it's a more of a matter of providing (and explaining) good default behaviours? and the tiny fraction of iOS OF projects that need more than simple playback or audio reactiveness will have to dig a bit into the sound classes to tweak them to taste?

**Comment by danoli3** [comment]
Need to respond to the errors and reconfigure on changes

**Comment by artificiel** [comment]
that's ideal, but as of right now ofSoundPlayer and ofSoundStream are mutually incompatible on iOS, due to their design/implementation.

my question is what does OF wants to cover? 

wrapping up the whole obj-c interface into OF C++ patterns correctly is not simple and requires to anticipate a lot of usage, which serves a very small subset of OF users. so my question is: does OF strive to cover all the possibilities (and properly document/support them) or provide some cross-platform solutions for common problems (as do ofSoundPlayer and ofSoundStream currently when used independently)?

until someone volunteers to go in deep and rewrite things, I suggest the solution to this issue is tweaking the examples so they work correctly, and add some documentation pointing to the subtleties of iOS routing and caveats on mixing ofSoundPlayer and ofSoundStream.

**Comment by ofTheo** [comment]
@danoli3 would be great if we could shed a bit more light on what might be a fix here?
I am guessing some of the audio fixes could already be in the AVEngine PR we were working on?
#6728 ( re-opened as #7293 ) 

In general it would be great to find some time to do an iOS call / sync up. I'd be happy to through some time/energy into fixing the simulator libs issue and this sound issue would be great to resolve. 

@artificiel there is a iOS working group here which I can tag you on if you are interested?
https://github.com/openframeworks/openFrameworks/discussions/7174

**Comment by artificiel** [comment]
@ofTheo I'll be working on an iOS project that includes sound (no input, just variable/poly playback) over the next few months and I will have pockets of availability to contribute.

I'm just now going over the AVEngine PR and trying to understand the scope/goal... it's not all clear to me what's done and what's missing (also re: ofxSoundObjects, which would be a fine integration) but for sure, if it's going to be interacting with `[AVAudioSession sharedInstance]` on iOS it turns this current issue in a moot point.
